### PR TITLE
Ensure backward compatibility with QueueType = Private/Public param

### PR DESCRIPTION
**Pull request by hadhh**
Hi,

Following our discussion on https://github.com/SNikalaichyk/cMsmq/pull/4, I made the changes to use a QueueTypeparam to ensure backward compatibility.

Notice however, that it prevents to have same name for public & private queue in the same config (as cMsmq use only Name as key instead of Name,QueueType but that the price to pay for 100 backward compatibility (can't have a default value on a key in dsc)

### CODE REVIEW

File: Examples/Sample_cMsmq.ps1 [hunk 1]
```diff
@@ -77,13 +81,35 @@ configuration Sample_cMsmq
...
+        DependsOn    = '[cMsmqQueue]PublicQueue1'
+    }
 }
 
 Sample_cMsmq -OutputPath "$Env:SystemDrive\Sample_cMsmq"
 
 Start-DscConfiguration -Path "$Env:SystemDrive\Sample_cMsmq" -Force -Verbose -Wait
 
-Get-DscConfiguration
+Get-DscConfiguration
```
Comment (SNikalaichyk) [inline Examples/Sample_cMsmq.ps1]: Please ensure there's an empty line at the end of file.

File: DSCResources/cMsmqQueuePermissionEntry/cMsmqQueuePermissionEntry.psm1 [hunk 2]
```diff
@@ -363,4 +402,38 @@ function Reset-cMsmqQueueSecurity
     }
 }
 
+function Build-QueuePath()
```
Comment (SNikalaichyk) [inline DSCResources/cMsmqQueuePermissionEntry/cMsmqQueuePermissionEntry.psm1]: Please use an [approved verb](https://msdn.microsoft.com/en-us/library/ms714428(v=vs.85).aspx) for this helper function.

File: DSCResources/cMsmqQueue/cMsmqQueue.psm1 [hunk 3]
```diff
@@ -71,6 +77,11 @@ function Test-TargetResource
         [String]
         $Name,
 
+        [Parameter(Mandatory = $false)] # not mandatory for backward compatibility (even if it prevents to have same name for public & private queue)
```
Comment (SNikalaichyk) [inline DSCResources/cMsmqQueue/cMsmqQueue.psm1]: I think this comment is not really needed. But if you insist on leaving it, please place it on a separate line.

File: DSCResources/cMsmqQueue/cMsmqQueue.psm1 [hunk 4]
```diff
@@ -207,6 +218,11 @@ function Set-TargetResource
...
```
Comment (SNikalaichyk) [inline DSCResources/cMsmqQueue/cMsmqQueue.psm1]: I think this comment is not really needed. But if you insist on leaving it, please place it on a separate line.

File: DSCResources/cMsmqQueue/cMsmqQueue.psm1 [hunk 5]
```diff
@@ -348,15 +365,21 @@ function Get-cMsmqQueue
...
         [ValidateNotNullOrEmpty()]
         [String]
-        $Name
+        $Name,
+
+        [Parameter(Mandatory = $false)]
+        [ValidateSet('Private', 'Public')]
+        [String]
+        $QueueType = 'Private'
+
```
Comment (SNikalaichyk) [inline DSCResources/cMsmqQueue/cMsmqQueue.psm1]: Please remove this empty line.

File: DSCResources/cMsmqQueue/cMsmqQueue.psm1 [hunk 6]
```diff
@@ -755,4 +820,39 @@ function Set-cMsmqQueue
...
+    )
+    if ($QueueType -eq 'Private')
+    {
+        $QueuePath = '{0}\{1}$\{2}' -f $env:COMPUTERNAME, $QueueType, $Name
+    }else{
```
Comment (SNikalaichyk) [inline DSCResources/cMsmqQueue/cMsmqQueue.psm1]: For consistency with the rest of the code, please place opening and closing braces on their own lines.

File: DSCResources/cMsmqQueuePermissionEntry/cMsmqQueuePermissionEntry.psm1 [hunk 7]
```diff
@@ -363,4 +402,38 @@ function Reset-cMsmqQueueSecurity
...
+        [Parameter(Mandatory = $false)]
+        [ValidateSet('Private', 'Public')]
+        [String]
+        $QueueType = 'Private'
+
+    )
+    if ($QueueType -eq 'Private')
+    {
+        $QueuePath = '{0}\{1}$\{2}' -f $env:COMPUTERNAME, $QueueType, $Name
+    }else{
```
Comment (SNikalaichyk) [inline DSCResources/cMsmqQueuePermissionEntry/cMsmqQueuePermissionEntry.psm1]: For consistency with the rest of the code, please place opening and closing braces on their own lines.

**Review by SNikalaichyk — CHANGES_REQUESTED** [review]
@hadhh, it looks really impressive, thanks! Could you address the following minor issues I found?

File: SourceFiles/cMsmq.cs [hunk 8]
```diff
@@ -254,7 +236,7 @@ private static GCHandle GetSecurityDescriptorHandle(QueuePath queuePath, int sec
             return gchSecurityDescriptor;
         }
 
-        public static MessageQueueAccessRights GetAccessMask(QueuePath queuePath, string userName)
+        public static MessageQueueAccessRights GetAccessMask(string queuePath, string userName)
```
Comment (SNikalaichyk) [inline SourceFiles/cMsmq.cs]: Please ensure there are no trailing spaces.

File: SourceFiles/cMsmq.cs [hunk 9]
```diff
@@ -266,7 +248,7 @@ public static MessageQueueAccessRights GetAccessMask(QueuePath queuePath, string
             return aceMask;
         }
 
-        public static string GetOwner(QueuePath queuePath)
+        public static string GetOwner(string queuePath)
```
Comment (SNikalaichyk) [inline SourceFiles/cMsmq.cs]: Same here, please ensure there are no trailing spaces.

**Comment by hadhh** [comment]
@SNikalaichyk  I reviewed the code according to your comments.

**Comment by hadhh** [comment]
@SNikalaichyk I also add a commit to fix an issue with the Get-TargetResource of PermissionEntry.

**Review by SNikalaichyk — APPROVED** [review]

**Comment by SNikalaichyk** [comment]
Merged, thanks!
I plan to add some Pester tests before I release the new version to the PS Gallery. It may take about a week as I don't have much free time nowadays.

**Pull request merged**